pkg/loop/internal/median.go

```diff
@@ -204,7 +205,7 @@ type reportCodecClient struct {
 }
 
 func (r *reportCodecClient) BuildReport(observations []median.ParsedAttributedObservation) (report libocr.Report, err error) {
-	ctx, cancel := r.ctx()
+	ctx, cancel := utils.ContextFromChan(r.StopCh)
```

Comment:
These still seem very noisy. What exactly was missing from `ctx()`? Would something like `stopCtx` say enough?
```suggestion
	ctx, cancel := r.stopCtx()
```

Reply:
How's `r.CtxFromStopCh()`? 

I'd like to read it and know it's not creating a new ctx from a parent context associated with the receiver. If we don't expect a lot of fresh eyes on this, though, we can go with the shorthand

Reply:
> I'd like to read it and know it's not creating a new ctx from a parent context associated with the receiver.

Why? That is an irrelevant implementation detail. The lifetime of the context is the only relevant information, not the mechanism of implementation.

Reply:
> The lifetime of the context is the only relevant information, not the mechanism of implementation.

Do you think the mechanism of implementation = lifetime of the context in this case? When `chStop` receives, the context is canceled. The context is passed through to the 3rd party grpc `Invoke` method, so the mechanism of implementation communicates that the calls to `Invoke` won't block even if stop channel receives. 
```
// ContextFromChan creates a context that finishes when the provided channel
// receives or is closed.
// When channel closes, the ctx.Err() will always be context.Canceled
// NOTE: Spins up a goroutine that exits on cancellation.
// REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
func ContextFromChan(chStop <-chan struct{}) (context.Context, context.CancelFunc) {
	ctx, cancel := context.WithCancel(context.Background())
	go func() {
		select {
		case <-chStop:
			cancel()
		case <-ctx.Done():
		}
	}()
	return ctx, cancel
}
```

Outside of renaming, is there a deeper thread to pull on wrt to creating and injecting this type of context into all grpc methods? i.e. 
```
func (s *serviceClient) Close() error {
	ctx, cancel := utils.ContextFromChan(s.b.StopCh)
	defer cancel()

	_, err := s.grpc.Close(ctx, &emptypb.Empty{})
	return err
}

func (s *serviceClient) Ready() error {
	ctx, cancel := utils.ContextFromChan(s.b.StopCh)
	defer cancel()
	ctx, cancel = context.WithTimeout(ctx, time.Second)
	defer cancel()

	_, err := s.grpc.Ready(ctx, &emptypb.Empty{})
	return err
}
```
Would it be feasible to have such a context initialized as field of `serviceClient`?
```
type serviceClient struct {
	b             *brokerExt
	cc           grpc.ClientConnInterface
	grpc       pb.ServiceClient
	stopCtx context.Context
}

func newServiceClient(b *brokerExt, cc grpc.ClientConnInterface) *serviceClient {
	ctx, _ := utils.ContextFromChan(b.StopCh)
	return &serviceClient{b, cc, pb.NewServiceClient(cc), ctx}
}

func (s *serviceClient) Start(ctx context.Context) error {
	return nil // no-op: server side starts automatically
}

func (s *serviceClient) Close() error {
	ctx, cancel := context.WithCancel(s.stopCtx)
	defer cancel()

	_, err := s.grpc.Close(ctx, &emptypb.Empty{})
	return err
}

func (s *serviceClient) Ready() error {
	ctx, cancel := context.WithCancel(s.stopCtx)
	defer cancel()

	ctx, cancel = context.WithTimeout(ctx, time.Second)
	defer cancel()

	_, err := s.grpc.Ready(ctx, &emptypb.Empty{})
	return err
}
```

Reply:
> Do you think the mechanism of implementation = lifetime of the context in this case?

No, because it could be implemented in other ways while exhibiting the exact same behavior for the caller.

> Would it be feasible to have such a context initialized as field of serviceClient?

We do often store context fields on structs instead, but it is an anti-pattern since contexts are "request" scoped.

Reply:
> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:

https://pkg.go.dev/context